File: docs/site/examples/python/consume.py
import asapo_consumer

endpoint = "localhost:8400"
beamtime = "asapo_test"

# test token. In production it is created during the start of the beamtime
token = str("eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.e"
"yJleHAiOjk1NzE3MTAyMTYsImp0aSI6ImMzaXFhbGpmNDNhbGZ"
"wOHJua20wIiwic3ViIjoiYnRfYXNhcG9fdGVzdCIsIkV4dHJhQ"
"2xhaW1zIjp7IkFjY2Vzc1R5cGVzIjpbIndyaXRlIiwicmVhZCJ"
"dfX0.dkWupPO-ysI4t-jtWiaElAzDyJF6T7hu_Wz_Au54mYU")

# set it  according to your configuration.
path_to_files = "/var/tmp/asapo/global_shared/data/test_facility/gpfs/test/2019/data/asapo_test"


consumer = asapo_consumer \
                .create_consumer(endpoint,
                                 path_to_files,
                                 True,           # True if the path_to_files is accessible locally, False otherwise
                                 beamtime,       # Same as for the producer
                                 "test_source",  # Same as for the producer
                                 token,          # Access token
                                 5000)           # Timeout. How long do you want to wait on non-finished stream for a message.


# you can get info about the streams in the beamtime
for stream in consumer.get_stream_list():
    print("Stream name: ", stream['name'], "\n",
          "LastId: ", stream['lastId'], "\n",
          "Stream finished: ", stream['finished'], "\n",
          "Next stream: ", stream['nextStream'])


group_id = consumer.generate_group_id() # Several consumers can use the same group_id to process messages in parallel

try:

    # get_next is the main function to get messages from streams. You would normally call it in loop.
    # you can either manually compare the meta['_id'] to the stream['lastId'], or wait for the exception to happen
    while True:
        data, meta = consumer.get_next(group_id, meta_only = False)
        print(data.tobytes().decode("utf-8"), meta)

except asapo_consumer.AsapoStreamFinishedError:
    print('stream finished') # all the messages in the stream were processed
        
except asapo_consumer.AsapoEndOfStreamError:
    print('stream ended')    # not-finished stream timeout, or wrong or empty stream

consumer.delete_stream(error_on_not_exist = True) # you can delete the stream after consuming

File: docs/site/examples/c/consume.c
#include "asapo/consumer_c.h"

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

void exit_if_error(const char *error_string, const AsapoErrorHandle err) {
    if (asapo_is_error(err)) {
        char buf[1024];
        asapo_error_explain(err, buf, sizeof(buf));
        printf("%s %s\n", error_string, buf);
        exit(EXIT_FAILURE);
    }
}

int main(int argc, char* argv[]) {
    AsapoErrorHandle err = asapo_new_handle();
    AsapoMessageMetaHandle mm = asapo_new_handle();
    AsapoMessageDataHandle data = asapo_new_handle();

    const char *endpoint = "localhost:8400";
    const char *beamtime = "asapo_test";
    const char *token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjk1NzE3MTAyMTYsImp0aSI6ImMzaXFhbGpmNDNhbGZwOHJua20wIiwic3ViIjoiYnRfYXNhcG9fdGVzdCIsIkV4dHJhQ2xhaW1zIjp7IkFjY2Vzc1R5cGVzIjpbIndyaXRlIiwicmVhZCJdfX0.dkWupPO-ysI4t-jtWiaElAzDyJF6T7hu_Wz_Au54mYU";

    const char * path_to_files = "/var/tmp/asapo/global_shared/data/test_facility/gpfs/test/2019/data/asapo_test"; //set it according to your configuration.

    AsapoSourceCredentialsHandle cred = asapo_create_source_credentials(kProcessed,
                                                                        beamtime,
                                                                        "", "test_source", token);
    AsapoConsumerHandle consumer = asapo_create_consumer(endpoint,
                                                         path_to_files, 1,
                                                         cred,
                                                         &err);
    asapo_free_handle(&cred);

    exit_if_error("Cannot create consumer", err);
    asapo_consumer_set_timeout(consumer, 5000ull);

    AsapoStringHandle group_id = asapo_consumer_generate_new_group_id(consumer, &err);
    exit_if_error("Cannot create group id", err);

    asapo_consumer_get_next(consumer, group_id, &mm, &data, "default",&err);
    exit_if_error("Cannot get next record", err);

    printf("id: %llu\n", (unsigned long long)asapo_message_meta_get_id(mm));
    printf("file name: %s\n", asapo_message_meta_get_name(mm));
    printf("file content: %s\n", asapo_message_data_get_as_chars(data));


// delete stream
    asapo_consumer_delete_stream(consumer,"default", 1,1,&err);
    exit_if_error("Cannot delete stream", err);
    printf("stream deleted\n");

    asapo_free_handle(&err);
    asapo_free_handle(&mm);
    asapo_free_handle(&data);
    asapo_free_handle(&consumer);
    asapo_free_handle(&group_id);
    return EXIT_SUCCESS;
}

